File: java/src/main/java/io/github/gsoultan/anubis/Roles.java
package io.github.gsoultan.anubis;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;

/** The set of roles a caller holds. */
public final class Roles implements Iterable<Role> {
    private static final Roles EMPTY = new Roles(List.of());

    private final List<Role> items;

    private Roles(List<Role> items) {
        this.items = List.copyOf(items);
    }

    public static Roles empty() {
        return EMPTY;
    }

    public static Roles of(List<String> roles) {
        if (roles == null || roles.isEmpty()) {
            return EMPTY;
        }
        List<Role> out = new ArrayList<>(roles.size());
        for (String r : roles) {
            out.add(Role.of(r));
        }
        return new Roles(out);
    }

    public boolean has(Role role) {
        return items.contains(role);
    }

    public boolean has(String role) {
        return has(Role.of(role));
    }

    public boolean hasAny(String... roles) {
        for (String r : roles) {
            if (has(r)) {
                return true;
            }
        }
        return false;
    }

    /** Narrow to the roles one application defined. */
    public Roles ofApp(String app) {
        List<Role> out = new ArrayList<>();
        for (Role r : items) {
            if (r.app().equals(app)) {
                out.add(r);
            }
        }
        return new Roles(out);
    }

    public List<String> toStrings() {
        List<String> out = new ArrayList<>(items.size());
        for (Role r : items) {
            out.add(r.value());
        }
        return Collections.unmodifiableList(out);
    }

    public int size() {
        return items.size();
    }

    public boolean isEmpty() {
        return items.isEmpty();
    }

    @Override
    public Iterator<Role> iterator() {
        return items.iterator();
    }

    @Override
    public String toString() {
        return String.join(" ", toStrings());
    }
}

File: php/src/Roles.php
<?php

declare(strict_types=1);

namespace Anubis;

/**
 * The set of roles a caller holds.
 *
 * @implements \IteratorAggregate<int, Role>
 */
final readonly class Roles implements \IteratorAggregate, \Countable
{
    /** @var list<Role> */
    private array $items;

    /** @param iterable<Role|string> $roles */
    public function __construct(iterable $roles = [])
    {
        $out = [];
        foreach ($roles as $r) {
            $out[] = Role::of($r);
        }
        $this->items = $out;
    }

    public function has(Role|string $role): bool
    {
        $want = Role::of($role);
        foreach ($this->items as $r) {
            if ($r->equals($want)) {
                return true;
            }
        }
        return false;
    }

    public function hasAny(Role|string ...$roles): bool
    {
        foreach ($roles as $r) {
            if ($this->has($r)) {
                return true;
            }
        }
        return false;
    }

    /** Narrow to the roles one application defined. */
    public function ofApp(string $app): self
    {
        return new self(array_filter($this->items, static fn (Role $r): bool => $r->app() === $app));
    }

    /** @return list<string> */
    public function toStrings(): array
    {
        return array_map(static fn (Role $r): string => $r->value, $this->items);
    }

    public function getIterator(): \Traversable
    {
        return new \ArrayIterator($this->items);
    }

    public function count(): int
    {
        return count($this->items);
    }
}
